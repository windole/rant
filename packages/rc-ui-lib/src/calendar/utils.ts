/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
