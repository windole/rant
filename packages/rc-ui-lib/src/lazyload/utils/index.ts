import { ListenEvent } from '../PropsType';

/* eslint-disable prefer-rest-params */
const inBrowser = typeof window !== 'undefined';

export const DEFAULT_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

// 需要监听的事件
export const DEFAULT_EVENTS: ListenEvent[] = [
  'scroll',
  'wheel',
  'resize',
  'mousewheel',
  'animationend',
  'transitionend',
  'touchmove',
];

export const hasIntersectionObserver =
  inBrowser &&
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;

export const modeType = {
  event: 'event',
  observer: 'observer',
};

export const call = (fn): void => fn && fn();

export function setImage(
  element: HTMLMediaElement | HTMLIFrameElement | HTMLImageElement | HTMLElement,
  imagePath: string,
): void {
  if (
    element instanceof HTMLIFrameElement ||
    element instanceof HTMLImageElement ||
    element instanceof HTMLMediaElement
  ) {
    element.src = imagePath;
  } else {
    element.style.backgroundImage = `url('${imagePath}')`;
  }
}

export function loadImage(imagePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = imagePath;
    img.onload = () => resolve(imagePath);
    img.onerror = (err) => reject(err);
  });
}

export function addCssClassName(
  element: HTMLImageElement | HTMLElement,
  cssClassName: string,
): void {
  if (!element.className.includes(cssClassName)) {
    element.className += ` ${cssClassName}`;
  }
}

export function throttle(fn, threshold = 250, scope?) {
  let last;
  let deferTimer;
  return () => {
    // @ts-ignore
    const context = scope || this;

    const now = +new Date();
    const args = arguments;
    if (last && now < last + threshold) {
      // hold on to it
      clearTimeout(deferTimer);
      deferTimer = setTimeout(() => {
        last = now;
        fn.apply(context, args);
      }, threshold);
    } else {
      last = now;
      fn.apply(context, args);
    }
  };
}

export function debounce(func, wait, immediate?) {
  let timeout;
  let args;
  let context;
  let timestamp;
  let result;

  const later = function later() {
    const last = +new Date() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) {
          context = null;
          args = null;
        }
      }
    }
  };

  return function debounced() {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    context = this;
    args = arguments;
    timestamp = +new Date();

    const callNow = immediate && !timeout;
    if (!timeout) {
      timeout = setTimeout(later, wait);
    }

    if (callNow) {
      result = func.apply(context, args);
      context = null;
      args = null;
    }

    return result;
  };
}
