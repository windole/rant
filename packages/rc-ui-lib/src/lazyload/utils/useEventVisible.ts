import { useEffect, useRef, useState } from 'react';
import { debounce as debounceFunc, DEFAULT_EVENTS, throttle as throttleFunc } from '.';
import { EventOptions } from '../PropsType';

// try to handle passive events
let passiveEventSupported = false;
try {
  const opts = Object.defineProperty({}, 'passive', {
    get() {
      passiveEventSupported = true;
    },
  });
  window.addEventListener('test', null, opts);
  // eslint-disable-next-line no-empty
} catch (e) {}

const passiveEvent = passiveEventSupported ? { capture: false, passive: true } : false;

const useEventVisible = (
  eventOptions: EventOptions,
): [React.MutableRefObject<HTMLDivElement | HTMLImageElement>, boolean] => {
  const {
    listenEvents = DEFAULT_EVENTS,
    offset,
    scrollContainer,
    debounce,
    throttle,
  } = eventOptions;
  // 是否可见
  const [visible, setVisible] = useState(false);
  // // 包裹容器
  const containerRef = useRef<HTMLDivElement | HTMLImageElement>(null);

  useEffect(() => {
    // 检查函数
    let checkVisible = () => {
      // 如果容器不存在则不计算
      if (!containerRef.current) return;
      // 获取当前组件位置
      const { top, left, width, height } = containerRef.current?.getBoundingClientRect();
      // 获取滚动容器位置
      const {
        top: parentTop,
        left: parentLeft,
        width: parentWidth,
        height: parentHeight,
      } = scrollContainer.getBoundingClientRect();
      // 计算屏幕高度和宽度
      const windowInnerHeight = window.innerHeight || document.documentElement.clientHeight;
      const windowInnerWidth = window.innerWidth || document.documentElement.clientWidth;
      // 计算滚动容器和屏幕可视区的相交区域
      const intersectionTop = Math.min(parentTop, 0);
      const intersectionLeft = Math.min(parentLeft, 0);

      const intersectionHeight =
        Math.min(windowInnerHeight, parentTop + parentHeight) - intersectionTop;
      const intersectionWidth =
        Math.min(windowInnerWidth, parentLeft + parentWidth) - intersectionLeft;
      // 计算组件距离可视区的高度
      const offsetTop = top - intersectionTop;
      const offsetLeft = left - intersectionLeft;
      // 偏移量计算[top, left]
      const offsets = Array.isArray(offset) ? offset : [offset, offset];
      // 通过上述距离判断组件是否在可见区域
      const isVisible =
        offsetTop - offsets[0] <= intersectionHeight &&
        offsetTop + height + offsets[0] >= 0 &&
        offsetLeft - offsets[1] <= intersectionWidth &&
        offsetLeft + width + offsets[1] >= 0;

      setVisible(isVisible);
      // 如果可见取消当前组件的所有的监听
      if (isVisible) {
        listenEvents.forEach((event) => {
          scrollContainer?.removeEventListener(event, checkVisible);
        });
      }
    };

    // 使用防抖节流增强checkVisible函数，这里优先防抖，其次节流
    if (debounce) {
      // 防抖函数
      checkVisible = debounceFunc(checkVisible, debounce);
    } else if (throttle) {
      // 节流函数
      checkVisible = throttleFunc(checkVisible, throttle);
    }
    // 批量监听事件
    listenEvents.forEach((event) => {
      scrollContainer?.addEventListener(event, checkVisible, passiveEvent);
    });

    // 自动执行一次
    checkVisible();

    return () => {
      listenEvents.forEach((event) => {
        scrollContainer?.removeEventListener(event, checkVisible, passiveEvent);
      });
    };
  }, [scrollContainer]);
  return [containerRef, visible];
};

export default useEventVisible;
