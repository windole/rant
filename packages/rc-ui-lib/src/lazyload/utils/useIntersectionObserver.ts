import { useCallback, useEffect, useState } from 'react';

const useIntersectionObserver = (
  observerOptions: IntersectionObserverInit,
  handleIntersection: IntersectionObserverCallback,
): ((node: Element) => void) => {
  const [element, setElement] = useState<Element>(null);

  const handleElement = useCallback((node: Element) => {
    setElement(node);
  }, []);

  useEffect(() => {
    if (!element) return;
    const observer = new IntersectionObserver(handleIntersection, observerOptions);
    observer.observe(element);
    // eslint-disable-next-line consistent-return
    return () => observer && observer.disconnect && observer.disconnect();
  }, [handleIntersection, observerOptions, element]);
  return handleElement;
};

export default useIntersectionObserver;
