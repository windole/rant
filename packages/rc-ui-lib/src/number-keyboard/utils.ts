import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
