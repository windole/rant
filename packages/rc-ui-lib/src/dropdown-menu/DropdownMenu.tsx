import React, {
  useEffect,
  useContext,
  useRef,
  useState,
  useMemo,
  Context,
  createContext,
  forwardRef,
  useImperativeHandle,
} from 'react';
import classnames from 'classnames';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import { HAPTICS_FEEDBACK } from '../utils/constant';
import { isDef } from '../utils';
import useScrollParent from '../hooks/use-scroll-parent';
import useClickAway from '../hooks/use-click-away';
import useEventListener from '../hooks/use-event-listener';
import useRefs from '../hooks/use-refs';
import getRect from '../hooks/get-rect';
import {
  DropdownMenuProps,
  DropdownMenuContextProps,
  DropdownMenuInstance,
  DropdownItemInstance,
} from './PropsType';

export const DropdownMenuContext: Context<DropdownMenuContextProps> = createContext({});
const DropdownMenu = forwardRef<DropdownMenuInstance, DropdownMenuProps>((props, ref) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('dropdown-menu', prefixCls);
  const [titleFlag, setTitleFlag] = useState(props.shouldRenderTitle);
  const [showPopupIndex, setShowPopupIndex] = useState(null);
  const showPopupIndexRef = useRef<number>(null);
  const rect = useRef<{ bottom: number; top: number }>({ bottom: 0, top: 0 });
  const root = useRef<HTMLDivElement>();
  const barRef = useRef<HTMLDivElement>();
  const offsetRef = useRef<number>(0);
  const [childrenRefs, setChildrenRefs] = useRefs<DropdownItemInstance>();
  const scrollParent = useScrollParent(root);

  const opened = useMemo(() => {
    return showPopupIndex !== null;
  }, [showPopupIndex]);

  const barStyle = () => {
    if (opened && isDef(props.zIndex)) {
      return {
        zIndex: +props.zIndex + 1,
      };
    }
    return {};
  };

  const updateShowPopupIndex = (current: number) => {
    showPopupIndexRef.current = current;
    setShowPopupIndex(current);
  };

  const updateOffset = () => {
    if (barRef.current) {
      rect.current = getRect(barRef.current);
      if (props.direction === 'down') {
        offsetRef.current = rect.current.bottom;
      } else {
        offsetRef.current = window.innerHeight - rect.current.top;
      }
    }
  };

  const closeAll = () => {
    childrenRefs.forEach((item) => {
      item.toggle(false);
    });
    updateShowPopupIndex(null);
  };

  useImperativeHandle(ref, () => ({
    close: closeAll,
  }));

  const onClickAway = () => {
    if (props.closeOnClickOutside) {
      closeAll();
    }
  };

  useEffect(() => {
    if (barRef.current) {
      updateOffset();
    }
  }, [barRef.current]);

  const onScroll = () => {
    if (opened) {
      updateOffset();
    }
  };

  const showItem = (index: number) => {
    const item = childrenRefs[index];
    updateOffset();
    updateShowPopupIndex(index);
    item.toggle(true);
  };

  const onToggleItem = (active: number) => {
    childrenRefs.forEach((child: DropdownItemInstance, index: number) => {
      if (index === active) {
        // 且连续点击同一个dropdown-menu__bar时，显示or隐藏
        if (active === showPopupIndexRef.current) {
          updateShowPopupIndex(null);
          child.toggle();
          return;
        }
        showItem(active);
      } else {
        child.toggle(false, { immediate: false });
      }
    });
  };

  const renderTitle = (item, index: number) => {
    const showPopup = showPopupIndex === index;
    const { disabled, titleClass } = item;

    return (
      <div
        key={index}
        role="button"
        tabIndex={disabled ? -1 : 0}
        className={classnames(bem('item', { disabled }), { [HAPTICS_FEEDBACK]: !disabled })}
        onClick={() => {
          if (!disabled) {
            onToggleItem(index);
          }
        }}
      >
        <span
          className={classnames(
            bem('title', {
              down: showPopup === (props.direction === 'down'),
              active: showPopup,
            }),
            titleClass,
          )}
          style={{ color: showPopup ? props.activeColor : '' }}
        >
          <div className="rc-ellipsis">{item.renderTitle()}</div>
        </span>
      </div>
    );
  };

  useClickAway(root, onClickAway);
  useEventListener('scroll', onScroll, { target: scrollParent });

  const handleChange = (value) => {
    if (titleFlag) {
      setTitleFlag(false);
    } else {
      setTitleFlag(true);
    }
    updateShowPopupIndex(null);
    props.onChange?.(value);
  };

  return (
    <DropdownMenuContext.Provider
      value={{
        props,
        offset: offsetRef.current,
        onChange: handleChange,
        close: closeAll,
      }}
    >
      <div ref={root} className={classnames(bem(), props.className)} style={{ ...props.style }}>
        <div ref={barRef} style={barStyle()} className={classnames(bem('bar', { opened }))}>
          {childrenRefs.map(renderTitle)}
        </div>
        {React.Children.toArray(props.children)
          .filter(Boolean)
          .map((child: React.ReactElement, index: number) =>
            React.cloneElement(child, {
              ref: setChildrenRefs(index),
            }),
          )}
      </div>
    </DropdownMenuContext.Provider>
  );
});

DropdownMenu.defaultProps = {
  duration: 200,
  overlay: true,
  direction: 'down',
  zIndex: 2000,
  closeOnClickOutside: true,
  closeOnClickOverlay: true,
  shouldRenderTitle: false,
};

DropdownMenu.displayName = 'DropdownMenu';

export default DropdownMenu;
