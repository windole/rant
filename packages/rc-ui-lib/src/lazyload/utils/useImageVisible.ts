import { useCallback, useState } from 'react';
import { addCssClassName, loadImage, setImage } from '.';
import { IntersectionObserverEntryType, ObserverOptions } from '../PropsType';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useImageVisible = (
  observerOptions: ObserverOptions = defaultObserverOptions,
  errorImage: string,
): [(node: Element) => void, boolean] => {
  const [isLoaded, setIsLoaded] = useState(false);

  const handleIntersection = useCallback(
    (entries: IntersectionObserverEntryType[]) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLImageElement | HTMLElement;
          const url = target.dataset.src || '';
          loadImage(url)
            .catch(() => {
              setImage(target, errorImage);
              addCssClassName(target, 'rc-lazyload__error');
              target.setAttribute('lazy', 'error');
            })
            .then((imagePath: string) => {
              setImage(target, imagePath);
              setIsLoaded(true);
              addCssClassName(target, 'rc-lazyload__loaded');
              target.setAttribute('lazy', 'loaded');
              target.removeAttribute('data-src');
            });
        });
    },
    [setIsLoaded, isLoaded],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, isLoaded];
};

export default useImageVisible;
