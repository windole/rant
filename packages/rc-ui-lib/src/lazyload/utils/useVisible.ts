import { useCallback, useEffect, useState } from 'react';
import { addCssClassName } from '.';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useVisible = (
  initialVisible: boolean,
  observerOptions = defaultObserverOptions,
): [(node: Element) => void, boolean] => {
  const [visible, setVisible] = useState(initialVisible);

  const forceVisible = () => {
    setVisible(true);
  };

  useEffect(() => {
    if (initialVisible) {
      forceVisible();
    }
  }, [initialVisible]);

  const handleIntersection = useCallback(
    (entries) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLElement;
          addCssClassName(target, 'rc-lazyload__loaded');
          setVisible(true);
        });
    },
    [setVisible, visible],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, visible];
};

export default useVisible;
