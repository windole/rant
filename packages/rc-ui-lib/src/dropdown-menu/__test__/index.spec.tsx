import React from 'react';
import { fireEvent, render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import DropdownMenu from '../index';
import { sleep } from '../../../tests/utils';

const option1 = [
  { text: '全部商品', value: 0 },
  { text: '新款商品', value: 1 },
  { text: '活动商品', value: 2 },
];
const option2 = [
  { text: '默认排序', value: 'a' },
  { text: '好评排序', value: 'b' },
  { text: '销量排序', value: 'c' },
];

describe('DropdownMenu', () => {
  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  it('should render correctly when prop activeColor is set', async () => {
    const { container } = render(
      <DropdownMenu activeColor="#ff8100">
        <DropdownMenu.Item name="item1" options={option1} />
      </DropdownMenu>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when prop zIndex is set', async () => {
    const { container } = render(
      <DropdownMenu zIndex="3000">
        <DropdownMenu.Item name="item1" options={option1} />
      </DropdownMenu>,
    );
    expect(container).toMatchSnapshot();
  });

  it('closeOnClickOutside ', async () => {
    const { container } = render(
      <DropdownMenu>
        <DropdownMenu.Item name="item1" options={option1} />
      </DropdownMenu>,
    );
    await sleep();
    const title = container.querySelector('.rc-dropdown-menu__title');
    fireEvent.click(title);
    fireEvent.click(document.body);
    await sleep(10);
    expect(container).toMatchSnapshot();
  });

  it('direction up', async () => {
    const { container } = render(
      <DropdownMenu direction="up">
        <DropdownMenu.Item name="item1" options={option1} />
      </DropdownMenu>,
    );
    await sleep();
    expect(container).toMatchSnapshot();

    const titles = container.querySelectorAll('.rc-dropdown-menu__title');
    await fireEvent.click(titles[0]);

    expect(container).toMatchSnapshot();
  });

  it('zIndex', async () => {
    const { container, getByRole } = render(
      <DropdownMenu zIndex="3000">
        <DropdownMenu.Item name="item1" options={option1} />
      </DropdownMenu>,
    );

    const btn = getByRole('button');
    await fireEvent.click(btn);
    expect(container.querySelector('.rc-dropdown-menu__bar')).toHaveStyle('z-index:3001');
  });

  it('shouldRenderTitle', async () => {
    const onChange = jest.fn();
    const { container, getAllByRole } = render(
      <DropdownMenu shouldRenderTitle>
        <DropdownMenu.Item name="item1" options={option1} onChange={onChange} />
        <DropdownMenu.Item name="item2" options={option2} />
      </DropdownMenu>,
    );
    const btns = getAllByRole('button');
    await fireEvent.click(btns[0]);
    const options = container.querySelectorAll('.rc-dropdown-item__option');
    await fireEvent.click(options[0]);
    await fireEvent.click(btns[1]);
  });

  it('disable dropdown item', async () => {
    const { container, getByRole } = render(
      <DropdownMenu>
        <DropdownMenu.Item name="item1" options={option1} disabled />
      </DropdownMenu>,
    );
    await sleep();
    const btn = getByRole('button');
    await fireEvent.click(btn);
    expect(container.querySelector('.rc-dropdown-item')).toHaveStyle('display:none');
  });

  it('emit change event ', async () => {
    const onChange = jest.fn();
    const { container, getByRole } = render(
      <DropdownMenu>
        <DropdownMenu.Item name="item1" options={option1} onChange={onChange} />
      </DropdownMenu>,
    );
    await fireEvent.click(getByRole('button'));
    const options = container.querySelectorAll('.rc-dropdown-item__option');
    await fireEvent.click(options[0]);
    expect(onChange).toHaveBeenCalledWith({ text: '全部商品', value: 0 });
  });

  it('click on the same button twice ', async () => {
    const { container, getByRole } = render(
      <DropdownMenu>
        <DropdownMenu.Item name="item1" options={option1} />
      </DropdownMenu>,
    );
    const btn = getByRole('button');
    await fireEvent.click(btn);
    const item = container.querySelector('.rc-dropdown-item');
    expect(item).toHaveStyle('display:block');
    await fireEvent.click(btn);
    await sleep(500);
    expect(item).toHaveStyle('display:none');
  });

  it('click on the different buttons separately ', async () => {
    const { container, getAllByRole } = render(
      <DropdownMenu>
        <DropdownMenu.Item name="item1" options={option1} />
        <DropdownMenu.Item name="item2" options={option2} />
      </DropdownMenu>,
    );
    const btns = getAllByRole('button');
    await fireEvent.click(btns[0]);
    const items = container.querySelectorAll('.rc-dropdown-item');
    expect(items[0]).toHaveStyle('display:block');
    await fireEvent.click(btns[1]);
    await sleep(500);
    expect(items[0]).toHaveStyle('display:none');
  });

  it('title', async () => {
    const { container } = render(
      <DropdownMenu>
        <DropdownMenu.Item name="item1" title="please click" />
      </DropdownMenu>,
    );
    expect(container.querySelector('.rc-ellipsis')).toHaveTextContent('please click');
  });

  it('teleport', async () => {
    const { container, getByRole } = render(
      <DropdownMenu>
        <DropdownMenu.Item name="item1" options={option1} teleport={document.body} />
      </DropdownMenu>,
    );
    const btn = getByRole('button');
    await fireEvent.click(btn);
    const items = document.body.querySelectorAll('.rc-dropdown-item__option');
    fireEvent.click(items[0]);
    expect(container).toMatchSnapshot();
  });

  it('scroll', async () => {
    const { container, getByRole } = render(
      <DropdownMenu>
        <DropdownMenu.Item name="item1" options={option1} />
      </DropdownMenu>,
    );
    const btn = getByRole('button');
    await fireEvent.click(btn);

    fireEvent.scroll(document.body);

    expect(container).toMatchSnapshot();
  });
});
