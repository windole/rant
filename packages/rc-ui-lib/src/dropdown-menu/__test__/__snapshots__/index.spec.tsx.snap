// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu closeOnClickOutside  1`] = `
<div>
  <div
    class="rc-dropdown-menu"
  >
    <div
      class="rc-dropdown-menu__bar"
      style=""
    >
      <div
        class="rc-dropdown-menu__item rc-haptics-feedback"
        role="button"
        tabindex="0"
      >
        <span
          class="rc-dropdown-menu__title"
        >
          <div
            class="rc-ellipsis"
          >
            请选择
          </div>
        </span>
      </div>
    </div>
    <div
      class="rc-dropdown-item rc-dropdown-item--down"
      style="z-index: 2000; top: 0px; display: block;"
    >
      <div
        class="rc-overlay rc-fade-exit rc-fade-exit-active"
        style="z-index: 2000; position: absolute; animation-duration: 200ms;"
      />
      <div
        class="rc-popup rc-popup--top rc-dropdown-item__content rc-popup-slide-top-exit rc-popup-slide-top-exit-active"
        style="z-index: 2000; transition-duration: 200ms;"
      >
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            全部商品
          </div>
        </div>
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            新款商品
          </div>
        </div>
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            活动商品
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropdownMenu direction up 1`] = `
<div>
  <div
    class="rc-dropdown-menu"
  >
    <div
      class="rc-dropdown-menu__bar"
    >
      <div
        class="rc-dropdown-menu__item rc-haptics-feedback"
        role="button"
        tabindex="0"
      >
        <span
          class="rc-dropdown-menu__title rc-dropdown-menu__title--down"
        >
          <div
            class="rc-ellipsis"
          >
            请选择
          </div>
        </span>
      </div>
    </div>
    <div
      class="rc-dropdown-item rc-dropdown-item--up"
      style="z-index: 2000; bottom: 768px; display: none;"
    />
  </div>
</div>
`;

exports[`DropdownMenu direction up 2`] = `
<div>
  <div
    class="rc-dropdown-menu"
  >
    <div
      class="rc-dropdown-menu__bar rc-dropdown-menu__bar--opened"
      style="z-index: 2001;"
    >
      <div
        class="rc-dropdown-menu__item rc-haptics-feedback"
        role="button"
        tabindex="0"
      >
        <span
          class="rc-dropdown-menu__title rc-dropdown-menu__title--active"
        >
          <div
            class="rc-ellipsis"
          >
            请选择
          </div>
        </span>
      </div>
    </div>
    <div
      class="rc-dropdown-item rc-dropdown-item--up"
      style="z-index: 2000; bottom: 768px; display: block;"
    >
      <div
        class="rc-overlay rc-fade-enter rc-fade-enter-active"
        style="z-index: 2001; position: absolute; animation-duration: 200ms;"
      />
      <div
        class="rc-popup rc-popup--bottom rc-dropdown-item__content rc-popup-slide-bottom-enter rc-popup-slide-bottom-enter-active"
        style="z-index: 2001; transition-duration: 200ms;"
      >
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            全部商品
          </div>
        </div>
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            新款商品
          </div>
        </div>
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            活动商品
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropdownMenu scroll 1`] = `
<div>
  <div
    class="rc-dropdown-menu"
  >
    <div
      class="rc-dropdown-menu__bar rc-dropdown-menu__bar--opened"
      style="z-index: 2001;"
    >
      <div
        class="rc-dropdown-menu__item rc-haptics-feedback"
        role="button"
        tabindex="0"
      >
        <span
          class="rc-dropdown-menu__title rc-dropdown-menu__title--down rc-dropdown-menu__title--active"
        >
          <div
            class="rc-ellipsis"
          >
            请选择
          </div>
        </span>
      </div>
    </div>
    <div
      class="rc-dropdown-item rc-dropdown-item--down"
      style="z-index: 2000; top: 0px; display: block;"
    >
      <div
        class="rc-overlay rc-fade-enter rc-fade-enter-active"
        style="z-index: 2010; position: absolute; animation-duration: 200ms;"
      />
      <div
        class="rc-popup rc-popup--top rc-dropdown-item__content rc-popup-slide-top-enter rc-popup-slide-top-enter-active"
        style="z-index: 2010; transition-duration: 200ms;"
      >
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            全部商品
          </div>
        </div>
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            新款商品
          </div>
        </div>
        <div
          class="rc-cell rc-cell--clickable rc-dropdown-item__option"
          role="button"
        >
          <div
            class="rc-cell__title"
          >
            活动商品
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropdownMenu should render correctly when prop activeColor is set 1`] = `
<div>
  <div
    class="rc-dropdown-menu"
  >
    <div
      class="rc-dropdown-menu__bar"
    >
      <div
        class="rc-dropdown-menu__item rc-haptics-feedback"
        role="button"
        tabindex="0"
      >
        <span
          class="rc-dropdown-menu__title"
        >
          <div
            class="rc-ellipsis"
          >
            请选择
          </div>
        </span>
      </div>
    </div>
    <div
      class="rc-dropdown-item rc-dropdown-item--down"
      style="z-index: 2000; top: 0px; display: none;"
    />
  </div>
</div>
`;

exports[`DropdownMenu should render correctly when prop zIndex is set 1`] = `
<div>
  <div
    class="rc-dropdown-menu"
  >
    <div
      class="rc-dropdown-menu__bar"
    >
      <div
        class="rc-dropdown-menu__item rc-haptics-feedback"
        role="button"
        tabindex="0"
      >
        <span
          class="rc-dropdown-menu__title"
        >
          <div
            class="rc-ellipsis"
          >
            请选择
          </div>
        </span>
      </div>
    </div>
    <div
      class="rc-dropdown-item rc-dropdown-item--down"
      style="z-index: 3000; top: 0px; display: none;"
    />
  </div>
</div>
`;

exports[`DropdownMenu teleport 1`] = `
<div>
  <div
    class="rc-dropdown-menu"
  >
    <div
      class="rc-dropdown-menu__bar"
      style=""
    >
      <div
        class="rc-dropdown-menu__item rc-haptics-feedback"
        role="button"
        tabindex="0"
      >
        <span
          class="rc-dropdown-menu__title"
        >
          <div
            class="rc-ellipsis"
          >
            全部商品
          </div>
        </span>
      </div>
    </div>
  </div>
</div>
`;
